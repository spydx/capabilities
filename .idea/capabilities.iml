<?xml version="1.0" encoding="UTF-8"?>
<module type="CPP_MODULE" version="4">
  <component name="NewModuleRootManager">
    <content url="file://$MODULE_DIR$">
      <sourceFolder url="file://$MODULE_DIR$/capabilities-derive/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/target/debug/build/async-trait-5b223cdf2dc4e051/out" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/target/debug/build/proc-macro2-cecd7ae0bcf8d632/out" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/target/debug/build/syn-12d88766fd7433d1/out" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/tests" isTestSource="true" />
      <excludeFolder url="file://$MODULE_DIR$/target" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
  </component>
</module>