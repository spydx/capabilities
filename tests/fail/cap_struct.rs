use capabilities::SqliteDb;
use capabilities_derive::capabilities;
use capabilities_derive::service;
use capabilities::*;

#[capabilities(Create, Read, Update, Delete)]
struct Orders {
    id: i32,
    name: String,
}

#[service(SqliteDb)]
#[tokio::main]
async fn main() -> Result<(), std::io::Error> {
    let connection_string = "sqlite::memory:".to_string();
    let _pool = CapService::build(connection_string)
        .await
        .expect("Failed to create database");
    Ok(())
}
