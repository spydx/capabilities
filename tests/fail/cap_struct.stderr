error[E0252]: the name `Pool` is defined multiple times
  --> tests/fail/cap_struct.rs:12:1
   |
4  | use sqlx::Pool;
   |     ---------- previous import of the type `Pool` here
...
12 | #[service(PoolSqlite)]
   | ^^^^^^^^^^^^^^^^^^^^^^ `Pool` reimported here
   |
   = note: `Pool` must be defined only once in the type namespace of this module
   = note: this error originates in the attribute macro `service` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `CapService: Capability<capabilities::Delete<Orders>>` is not satisfied
 --> tests/fail/cap_struct.rs:6:38
  |
6 | #[capabilities(Create, Read, Update, Delete)]
  |                                      ^^^^^^ the trait `Capability<capabilities::Delete<Orders>>` is not implemented for `CapService`
  |
note: required by a bound in `CapDeleteOrders`
 --> tests/fail/cap_struct.rs:6:1
  |
6 | #[capabilities(Create, Read, Update, Delete)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `CapDeleteOrders`
  = note: this error originates in the macro `cap` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `CapService: Capability<capabilities::Update<Orders>>` is not satisfied
 --> tests/fail/cap_struct.rs:6:30
  |
6 | #[capabilities(Create, Read, Update, Delete)]
  |                              ^^^^^^ the trait `Capability<capabilities::Update<Orders>>` is not implemented for `CapService`
  |
note: required by a bound in `CapUpdateOrders`
 --> tests/fail/cap_struct.rs:6:1
  |
6 | #[capabilities(Create, Read, Update, Delete)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `CapUpdateOrders`
  = note: this error originates in the macro `cap` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `CapService: Capability<capabilities::Read<Orders>>` is not satisfied
 --> tests/fail/cap_struct.rs:6:24
  |
6 | #[capabilities(Create, Read, Update, Delete)]
  |                        ^^^^ the trait `Capability<capabilities::Read<Orders>>` is not implemented for `CapService`
  |
note: required by a bound in `CapReadOrders`
 --> tests/fail/cap_struct.rs:6:1
  |
6 | #[capabilities(Create, Read, Update, Delete)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `CapReadOrders`
  = note: this error originates in the macro `cap` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `CapService: Capability<capabilities::Create<Orders>>` is not satisfied
 --> tests/fail/cap_struct.rs:6:16
  |
6 | #[capabilities(Create, Read, Update, Delete)]
  |                ^^^^^^ the trait `Capability<capabilities::Create<Orders>>` is not implemented for `CapService`
  |
note: required by a bound in `CapCreateOrders`
 --> tests/fail/cap_struct.rs:6:1
  |
6 | #[capabilities(Create, Read, Update, Delete)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `CapCreateOrders`
  = note: this error originates in the macro `cap` (in Nightly builds, run with -Z macro-backtrace for more info)
